contrib/coredns/policy/policy.go

```diff
@@ -119,11 +120,39 @@ func (p *PolicyMiddleware) getEDNS0Attrs(r *dns.Msg) ([]*pb.Attribute, bool) {
 	return attrs, foundSourceIP
 }
 
+// This function validates the middleware response
+// Inputs is *PolicyMiddle, *Response and error
+// Output is DNS status and error
+func ValidateMiddlewareResponse(p *PolicyMiddleware, w dns.ResponseWriter, r *dns.Msg,response *Response, status int, err error) (int, error) {
+
+	if err != nil {
+		log.Printf("[ERROR] Policy validation failed due to error %s\n", err)
+		return dns.RcodeServerFailure, err
+	}
+
+	if response.Permit == true && response.Redirect == nil {
+		log.Printf("[INFO] Permitting request")
+		return status, nil
```

Comment:
Based on the description below, here is where you would have to do w.WriteMsg(upstreamResult).